### PR TITLE
ENH PHP 8.4 support

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/.github/issues/331

Do not squash merge, as there are multiple commits

### CODE REVIEW

File: tests/php/Dev/BacktraceTest.php [hunk 1]
```diff
@@ -161,7 +161,7 @@ public static function matchesFilterableClassProvider(): array
     #[DataProvider('matchesFilterableClassProvider')]
     public function testMatchesFilterableClass(string $className, string $filterableClass, bool $expected, string $message): void
     {
-        $reflectionMethod = new ReflectionMethod(Backtrace::class . '::matchesFilterableClass');
+        $reflectionMethod = new ReflectionMethod(Backtrace::class, 'matchesFilterableClass');
```
Comment (emteknetnz) [inline tests/php/Dev/BacktraceTest.php:164]: Fixed a deprecation about passing a single parameter to the constructor

Comment (emteknetnz) [inline tests/php/Dev/BacktraceTest.php:164]: Fixed deprecation warning about passing single parameter to the constuctuor

File: tests/php/Dev/CsvBulkLoaderTest.php [hunk 2]
```diff
@@ -57,8 +57,8 @@ public function testLoad()
         $filepath = $this->csvPath . 'PlayersWithHeader.csv';
         $file = fopen($filepath ?? '', 'r');
         $compareCount = $this->getLineCount($file);
-        fgetcsv($file); // pop header row
-        $compareRow = fgetcsv($file);
+        fgetcsv($file, escape: "\\"); // pop header row
```
Comment (emteknetnz) [inline tests/php/Dev/CsvBulkLoaderTest.php:60]: Fixes Deprecated: fgetcsv(): the $escape parameter must be provided as its default value will change

File: src/Dev/Deprecation.php [hunk 3]
```diff
@@ -169,10 +169,14 @@ private static function get_called_from_trace(array $backtrace, int $level): arr
     {
         $newLevel = $level;
         // handle closures inside withSuppressedNotice()
-        if (Deprecation::$insideNoticeSuppression
-            && substr($backtrace[$newLevel]['function'], -strlen('{closure}')) === '{closure}'
-        ) {
-            $newLevel = $newLevel + 2;
+        if (Deprecation::$insideNoticeSuppression) {
```
Comment (emteknetnz) [inline src/Dev/Deprecation.php:172]: PHP 8.4 changed how it displays closure strings, this is technically a bugfix, separate commit message reflects this

**Comment by GuySartorelli** [comment]
@emteknetnz There's still Ci failures on a PHP 8.4 run here

File: tests/php/Dev/SapphireTestTest.php [hunk 4]
```diff
@@ -271,11 +270,8 @@ public function testEnableErrorHandler(int $errno, ?string $expectedClass): void
     public static function provideEnableErrorHandler(): array
     {
         // Only E_USER_* errors can be triggered, so that's all that's being tested
+        // As of PHP 8.4, E_USER_ERROR can no longer be triggered without a PHP deprecation notice
         return [
-            'error' => [
-                'errno' => E_USER_ERROR,
-                'expectedClass' => ExpectedErrorException::class,
-            ],
```
Comment (GuySartorelli) [inline tests/php/Dev/SapphireTestTest.php:278]: This wasn't mentioned in the release notes but is in the docs: https://www.php.net/manual/en/errorfunc.constants.php#constant.e-user-error
> Usage of this constant with [trigger_error()](https://www.php.net/manual/en/function.trigger-error.php) is deprecated as of PHP 8.4.0. It is recommended to either [throw](https://www.php.net/manual/en/language.exceptions.php) an [Exception](https://www.php.net/manual/en/class.exception.php) or call [exit()](https://www.php.net/manual/en/function.exit.php) instead.

**Review by GuySartorelli — APPROVED** [review]
LGTM

**Pull request merged**